Sofa/framework/Core/src/sofa/core/objectmodel/VisitorDirection.h

```diff
@@ -0,0 +1,104 @@
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/core/config.h>
+
+namespace sofa::core::objectmodel
+{
+
+struct DirectionalVisitor
+{
+    virtual ~DirectionalVisitor() = default;
+    DirectionalVisitor() = default;
+    virtual void operator()(sofa::core::behavior::OdeSolver*) const {}
+    virtual void operator()(sofa::core::behavior::ConstraintSolver*) const {}
+    virtual void operator()(sofa::core::BaseMapping*) const {}
+    virtual void operator()(sofa::core::behavior::BaseMechanicalState*) const {}
+    virtual void operator()(sofa::core::behavior::BaseMass*) const {}
+    virtual void operator()(sofa::core::behavior::BaseForceField*) const {}
+    virtual void operator()(sofa::core::behavior::BaseInteractionForceField* force) const
+    {
+        this->operator()((behavior::BaseForceField*)force);
+    }
+    virtual void operator()(sofa::core::behavior::BaseProjectiveConstraintSet*) const {}
+    virtual void operator()(sofa::core::behavior::BaseConstraintSet*) const {}
+    virtual void operator()(sofa::core::behavior::BaseInteractionProjectiveConstraintSet* constraint) const
+    {
+        this->operator()((behavior::BaseProjectiveConstraintSet*)constraint);
+    }
+    virtual void operator()(sofa::core::behavior::BaseInteractionConstraint* constraint) const
+    {
+        this->operator()((behavior::BaseConstraintSet*)constraint);
+    }
+};
+
+struct SOFA_CORE_API TopDownVisitor : DirectionalVisitor
+{
+    TopDownVisitor() = default;
+};
+struct SOFA_CORE_API BottomUpVisitor : DirectionalVisitor
+{
+    BottomUpVisitor() = default;
+};
+
+inline TopDownVisitor topDownVisitor {};
```

Comment:
using the same name with only no capital letter at the beginning is misleading. Given the fact that this is kind of equivalent to a static object, why not add a suffix ?

Reply:
There is no guidelines in SOFA relative to the naming of global variables (I already saw a convention with a `g_` prefix). But I agree with you the names of the type and the variable are too similar. The STL has the same design for execution policy, and the variable names are truncated to differentiate with the type names. But here, I don't like it so much. Let's discuss that in the dev meeting.
Keep in mind that this is a detail of one of the implementation of the visitors. We still need to decide which one of both implementations we keep. That's the main question of this PR.